feat(docs): Add Working With Images in Markdown page (#15296)
gatsbyjs/gatsby: docs/docs/working-with-images-in-markdown.md

---
title: Working With Images in Markdown
---

When building Gatsby sites composed primarily of markdown pages, insertion of images can enhance the content around it. Adding images can be done in multiple ways.

## Featured images with Frontmatter fields

In sites like a blog, you may want to include a featured image that appears at the top of a page. One way to do this can be by grabbing the image filename from a frontmatter field and then transforming it with `gatsby-plugin-sharp` in a GraphQL query.

> It is assumed you already have programatically generated pages from markdown using renderers like `gatsby-transformer-remark` or `gatsby-plugin-mdx`. If not, take a read through up to [Part 7 of the Gatsby Tutorial](tutorial/part-seven/). This will build upon the tutorial and as such, `gatsby-transformer-remark` will be used for this example

Reviewer: ```suggestion
This solution assumes you already have programmatically generated pages from Markdown with renderers like `gatsby-transformer-remark` or `gatsby-plugin-mdx`. If not, take a read through up to [Part 7 of the Gatsby Tutorial](tutorial/part-seven/). This will build upon the tutorial and as such, `gatsby-transformer-remark` will be used for this example.
```

> Note: This can be done similarly using [MDX](/docs/mdx/) as well. Instead of the `markdownRemark` nodes in GraphQL, `Mdx` can be swapped in and should work.

To start out, download the plugins for Gatsby-image as mentioned in [Using gatsby-image](/docs/using-gatsby-image/).

```shell
npm install --save gatsby-image gatsby-transformer-sharp gatsby-plugin-sharp
```

You will also want to have `gatsby-source-filesystem` installed. Then, configure the various plugins in the `gatsby-config` file.

```js:title=gatsby-config.js
module.exports = {
  plugins: [
    `gatsby-plugin-sharp`,
    `gatsby-transformer-sharp`,
    `gatsby-transformer-remark`,
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        path: `${__dirname}/src/pages`,
      },
    },
  ],
}
```

Then, in an example markdown file, add a field called `featuredImage`:

```md:title=src/pages/example-post.md
---
title: My Amazing Post
featuredImage: ./awesome-image.png
---

Content goes here!
```

> The field is going to be the relative path to the image we want to use.

Reviewer: ```suggestion
The `featuredImage` field should include the relative path to an image you want to use.
```

> Next, you can now query it in GraphQL. If the filepath points to an actual image, it will be transformed into a `File` node in GraphQL and then you can get the image data out of it by using the `childImageSharp` field.

Reviewer 2: Thanks for the PR! I personally think that Next, is not needed for this part. What do you think?

Reviewer: Agreed!

> Next, you can now query it in GraphQL. If the filepath points to an actual image, it will be transformed into a `File` node in GraphQL and then you can get the image data out of it by using the `childImageSharp` field.

Reviewer: ```suggestion
You can now query for the featured image in GraphQL. If the filepath points to an actual image, it will be transformed into a `File` node in GraphQL and then you can get the image data out of it by using the `childImageSharp` field.
```

> This can be added to the page query in the markdown template file. In this example, a [Fluid query](/packages/gatsby-image/#fluid-queries) is used to make a responsive image.

Reviewer: ```suggestion
This can be added to the GraphQL query in a Markdown template file. In this example, a [Fluid query](/docs/gatsby-image#images-that-stretch-across-a-fluid-container) is used to make a responsive image.
```

```jsx:title=src/templates/blog-post.js
export const query = graphql`
  query PostQuery($slug: String!) {
    markdownRemark(fields: { slug: { eq: $slug } }) {
      html
      frontmatter {
        title
        // highlight-start
        featuredImage {
          childImageSharp {
            fluid(maxWidth: 800) {
              ...GatsbyImageSharpFluid
            }
          }
        }
        // highlight-end
      }
    }
  }
`
```

> Now in the markup for the page, load the `gatsby-image` package and pass the results of the graphQL query into an `<Img />` component.

Reviewer: ```suggestion
Also in the Markdown post template, import the `gatsby-image` package and pass the results of the graphQL query into an `<Img />` component.
```

```jsx:title=src/templates/blog-post.js
import React from "react"
import { graphql } from "gatsby"
import Layout from "../components/layout"
// highlight-start
import Img from "gatsby-image"
// highlight-end

export default ({ data }) => {
  let post = data.markdownRemark

  // highlight-start
  let featuredImgFluid = post.frontmatter.featuredImage.childImageSharp.fluid
  // highlight-end

  return (
    <Layout>
      <div>
        <h1>{post.frontmatter.title}</h1>
        // highlight-start
        <Img fluid={featuredImgFluid} />
        // highlight-end
        <div dangerouslySetInnerHTML={{ __html: post.html }} />
      </div>
    </Layout>
  )
}

export const query = graphql`
  query PostQuery($slug: String!) {
    markdownRemark(fields: { slug: { eq: $slug } }) {
      html
      frontmatter {
        title
        featuredImage {
          childImageSharp {
            fluid(maxWidth: 800) {
              ...GatsbyImageSharpFluid
            }
          }
        }
      }
    }
  }
`
```

Your featured image should now appear on the generated page right below the main header. Tada!

## Inline images with `gatsby-remark-images`

> Images also may be included in the markdown body itself. To do this, the plugin [gatsby-remark-images](/packages/gatsby-remark-images) comes in handy.

Reviewer: ```suggestion
Images also may be included in the Markdown body itself. The plugin [gatsby-remark-images](/packages/gatsby-remark-images) comes in handy for this.
```

Start out by installing `gatsby-remark-images` and `gatsby-plugin-sharp`.

```shell
npm install --save gatsby-remark-images gatsby-plugin-sharp
```

Also be make sure that `gatsby-source-filesystem` is installed and points at the directory where your images are located.

> Then, configure the plugins by adding it into your `gatsby-config` file. As with the previous example, either `Remark` or `MDX` can be used. `gatsby-plugin-mdx` will be used in this usecase. Put the `gatsby-remark-images` plugin within the `gatsbyRemarkPlugins` option field of `gatsby-plugin-mdx`.

Reviewer: ```suggestion
Configure the plugins in your `gatsby-config` file. As with the previous example, either `Remark` or `MDX` can be used; `gatsby-plugin-mdx` will be used in this case. Put the `gatsby-remark-images` plugin within the `gatsbyRemarkPlugins` option field of `gatsby-plugin-mdx`.
```

```js:title=gatsby-config.js
module.exports = {
  plugins: [
    `gatsby-plugin-sharp`,
    {
      resolve: `gatsby-plugin-mdx`,
      options: {
        gatsbyRemarkPlugins: [
          {
            resolve: `gatsby-remark-images`,
            options: {
              maxWidth: 1200,
            },
          },
        ],
      },
    },
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        path: `${__dirname}/src/pages`,
      },
    },
  ],
}
```

> Then, we can use the default markdown syntax for images and they will be processed by Sharp and will appear as if we placed it in a `gatsby-image` component.

Reviewer: ```suggestion
With this configuration, you can use the default Markdown syntax for images. They will be processed by Sharp and appear as if you placed them in a `gatsby-image` component.
```
